File: Cow-Checkups.cpp
// #include <bits/stdc++.h>
// using namespace std;

// int main() {
//     ios::sync_with_stdio(false);
//     cin.tie(nullptr);

//     int N;
//     cin >> N;

//     vector<int> cow(N), need(N);
//     for (int i = 0; i < N; i++) cin >> cow[i];
//     for (int i = 0; i < N; i++) cin >> need[i];

//     // Step 1: original matches and prefix sum
//     vector<int> base(N), pref(N + 1, 0);
//     for (int i = 0; i < N; i++) {
//         base[i] = (cow[i] == need[i]);
//         pref[i + 1] = pref[i] + base[i];
//     }

//     // Step 2: diagonal prefix sums
//     // diag[s][i] = number of matches cow[k] == need[s-k] for k in [0, i)
//     vector<vector<int>> diag(2 * N - 1, vector<int>(N + 1, 0));

//     for (int s = 0; s < 2 * N - 1; s++) {
//         for (int i = 0; i < N; i++) {
//             diag[s][i + 1] = diag[s][i];
//             int j = s - i;
//             if (0 <= j && j < N && cow[i] == need[j]) {
//                 diag[s][i + 1]++;
//             }
//         }
//     }

//     // Step 3: count all reversals
//     vector<long long> ans(N + 1, 0);

//     for (int l = 0; l < N; l++) {
//         for (int r = l; r < N; r++) {
//             int outside = pref[l] + (pref[N] - pref[r + 1]);
//             int s = l + r;
//             int inside = diag[s][r + 1] - diag[s][l];
//             int total = outside + inside;
//             ans[total]++;
//         }
//     }

//     // Output
//     for (int i = 0; i <= N; i++) {
//         cout << ans[i] << '\n';
//     }

//     return 0;
// }

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;

    vector<int> cow(N), need(N);
    for (int i = 0; i < N; i++) cin >> cow[i];
    for (int i = 0; i < N; i++) cin >> need[i];

    // prefix matches (same as Python)
    vector<int> pref(N + 1, 0);
    for (int i = 0; i < N; i++) {
        pref[i + 1] = pref[i] + (cow[i] == need[i]);
    }

    vector<long long> ans(N + 1, 0);
    vector<int> diag(N + 1);   // reused diagonal buffer

    // iterate diagonals in increasing s = l + r
    for (int s = 0; s < 2 * N - 1; s++) {

        // build diag[s][*]
        diag[0] = 0;
        for (int i = 0; i < N; i++) {
            diag[i + 1] = diag[i];
            int j = s - i;
            if (0 <= j && j < N && cow[i] == need[j]) {
                diag[i + 1]++;
            }
        }

        // EXACT same (l, r) logic, just grouped by s
        for (int l = max(0, s - (N - 1)); l <= min(N - 1, s); l++) {
            int r = s - l;
            if (l > r) continue;   // matches original l <= r condition

            int out = pref[l] + (pref[N] - pref[r + 1]);
            int ins = diag[r + 1] - diag[l];
            ans[out + ins]++;
        }
    }

    for (int i = 0; i <= N; i++) {
        cout << ans[i] << '\n';
    }

    return 0;
}

File: Cow-Checkups.py
# N=int(input())
# cow=list(map(int,input().split()))
# need=list(map(int,input().split()))

# start=[1 if cow[i]==need[i] else 0 for i in range(N)]
# pref=[0]*(N+1)
# for i in range(N):
#     pref[i+1]=pref[i]+start[i]


# #[s][i] is num match cow[k] is need[s-k] for k in [0, i)
# diag=[[0]*(N + 1) for _ in range(2*N-1)]

# for s in range(2*N-1):
#     for i in range(N):
#         diag[s][i+1]=diag[s][i]
#         j=s-i
#         if 0<=j<N and cow[i]==need[j]:
#             diag[s][i+1]+=1

# ans=[0]*(N+1)

# for l in range(N):
#     for r in range(l,N):
#         out=pref[l]+(pref[N]-pref[r+1])
#         s=l+r
#         ins=diag[s][r+1]-diag[s][l]
#         total=out+ins
#         ans[total]+=1

# for x in ans:
#     print(x)


import sys
input = sys.stdin.readline

N = int(input())
cow = list(map(int, input().split()))
need = list(map(int, input().split()))

# prefix matches
pref = [0] * (N + 1)
for i in range(N):
    pref[i + 1] = pref[i] + (cow[i] == need[i])

# cache for diagonals
diag_cache = {}

def get_diag(s):
    if s in diag_cache:
        return diag_cache[s]

    d = [0] * (N + 1)
    for i in range(N):
        d[i + 1] = d[i]
        j = s - i
        if 0 <= j < N and cow[i] == need[j]:
            d[i + 1] += 1

    diag_cache[s] = d
    return d

ans = [0] * (N + 1)

# EXACT same (l, r) logic as original
for l in range(N):
    for r in range(l, N):
        out = pref[l] + (pref[N] - pref[r + 1])
        s = l + r
        d = get_diag(s)
        ins = d[r + 1] - d[l]
        ans[out + ins] += 1

for x in ans:
    print(x)
